feat: add net_ldap instrumentation

scbjans opened the pull request:
Largely based on prior work in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1432 by @Hareramrai and @jdehaan

Since that PR seems stale, I've updated it with some small edits to comply with release management using toys.

Should fix: github.com/open-telemetry/opentelemetry-ruby-contrib/issues/669

scbjans commented on .toys/.data/releases.yml:
```diff
@@ -231,7 +237,8 @@ gems:
   - name: opentelemetry-propagator-google_cloud_trace_context
     directory: propagator/google_cloud_trace_context
-    version_constant: [OpenTelemetry, Propagator, GoogleCloudTraceContext, VERSION]
+    version_constant:
```
Thanks for your quick review @arielvalentin !

I guess this is from the `format on save` option in VSCode which I accidentally committed. I'll revert it.

I'll respond to the other comments this Friday.

arielvalentin commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/patches/instrumentation.rb:
```diff
@@ -0,0 +1,31 @@
+            def initialize(args = {})
+              super
+
+              @instrumentation_service = args[:instrumentation_service] || OpenTelemetry::Instrumentation::Net::LDAP::InstrumentationService.new({
```
Another thought I have here, what if we added this otel instrumentation as a decorator to the injected service?

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,84 @@
+            LDAP::Instrumentation.instance.config
+          end
+
+          def annotate_span_with_response!(span, response)
```
https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/0583992b7cf2d4262a48672efdc8b8885ab7b1e3

kaylareopelle commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,84 @@
+            tracer.in_span(
+              event,
+              attributes: attributes.compact,
+              kind: :client
```
Good points. I was also wondering about SemConv as I'm reading through this. I asked the [#opentelemetry-semantic-conventions](https://cloud-native.slack.com/archives/C041APFBYQP/p1758067382218469) channel on CNCF Slack for advice.

kaylareopelle replied:
Heard back from the thread. There aren't any specific LDAP conventions we should adhere to. Instead, they suggested we use attributes from other relevant categories like [network](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/registry/attributes/network.md) and [server](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/registry/attributes/server.md).

kaylareopelle reviewed (commented):
Hi @scbjans, thanks for opening this PR. 

I'm not very familiar with LDAP, so I'll take another pass once I learn a bit more. For now, I left a few comments.

In addition to those comments, would you mind updating the CODEOWNERS file to list you as the owner of this gem? PR assignment for CODEOWNERS is currently broken, but I'm hoping to get it fixed soon.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+              'ldap.base' => base,
+              'ldap.encryption' => encryption.to_json,
+              'ldap.payload' => payload.to_json,
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_NAME => host || hosts,
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_PORT => port,
```
Thanks! I see the new opentelemetry-semantic-conventions gem has been released yesterday, I'll update this PR accordingly

thompson-tomo replied:
Possibly both as per https://opentelemetry.io/docs/specs/semconv/general/attributes/#clientserver-examples-using-networkpeer

thompson-tomo replied:
Resolved.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+              message = ::Net::LDAP.result2string(status_code)
+            end
+            attributes = {
+              'ldap.status_code' => status_code
```
I like that, brings it more in line with the HTTP semconv

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,84 @@
+            @encryption = args[:encryption]
+          end
+
+          def instrument(event, payload)
```
I think a good starting point would be to capture the name of the operation being performed ie open, read, parse_pdu.

scbjans commented:
@kaylareopelle thanks for your feedback; I updated it in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/7264acb6e8dbfccb0c2e11c04d56b3984068a18c

@thompson-tomo thank you as well for your feedback, I believe I addressed it in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/44a1d5413a831ae8f3c325e57fd00ce4d592c569
However, tests are failing, because the new `opentelemetry-semantic_conventions` is missing some `require`s of the `SemConv` module. I'll have to tidy up my fixes for that and submit a PR

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+                annotate_span_with_response(span, response) if response
+              end
+            rescue ::Net::LDAP::Error => e
+              span.set_attribute('ldap.error_message', "#{e.class}: #{e.message}")
```
Resolved.

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_NAME => host || hosts,
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_PORT => port,
+              OpenTelemetry::SemanticConventions::Trace::PEER_SERVICE => instrumentation_config[:peer_service]
+            }
```
Resolved.

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+              'ldap.status_code' => status_code
+            }
+            attributes['ldap.message'] = message unless message.empty?
+            attributes['ldap.error_message'] = error_message unless error_message.empty?
```
Resolved.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,95 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry/semconv/incubating/peer'
```
I do not like this line; but I'll add a test to make sure we catch it when the constant is removed from `Incubating`

scbjans replied:
https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/111f18c81dd12f26cc0faa0d4599205148686638

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,88 @@
+
+          def instrument(event, payload)
+            attributes = {
+              'ldap.auth' => auth.except(:password).to_json,
```
I believe it was in relation to the range of different auth methods/mechanism that could be used ie certificates vs user credentials.

The name Auth doesn't feel scoped enough as it could mean many things. In ldap v3 SASL is supported which opens up additional techniques.

thompson-tomo replied:
If we retain it, then we should go with ldap.auth.method etc

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,95 @@
+          def instrument(event, payload)
+            attributes = {
+              'ldap.auth' => auth.except(:password).to_json,
+              'ldap.base' => base,
```
Should this become `ldap.tree.base` or similar that way we have potential of adding other properties for the tree.

scbjans replied:
Done so in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/dc069a9769e46b33198681b2a25df93ab3a0eb08

thompson-tomo commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,95 @@
+            attributes = {
+              'ldap.auth' => auth.except(:password).to_json,
+              'ldap.base' => base,
+              'ldap.encryption' => encryption.to_json,
```
Not sure of the benefits of this given it could be a hash of properties.

thompson-tomo replied:
Agree best to leave it out for now and can be added if need arises but use the tls namespace where appropriate.

scbjans replied:
Removed it in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/97837bc17ca55cc5a7cd5c4e7bcc44f2e10501b3

thompson-tomo reviewed (approved):
Span definition & attribute usage looks good especially as it is consistent with existing sem conv definitions. 👍 

Will allow others to validate the coding patterns/structure.

scbjans committed:
feat: add net_ldap instrumentation

scbjans committed:
chore: add net_ldap to component_owners

scbjans committed:
chore: do not include internal spans by default

scbjans committed:
chore: instrumentation should not break when searching for binary objectsids

scbjans committed:
chore: appease the cop

scbjans committed:
chore: do not add attribute when JSON could not be generated.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,112 @@
+              'ldap.request.message' => begin
+                payload.to_json
+              rescue JSON::GeneratorError
+                'Could not generate JSON'
```
After thinking about this, I think it's better to set it to `nil`, instead of changing the value type as well.
Done so in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/fe142ea760501550c8ac75c6d663ddb42a9c691b

arielvalentin commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/attribute_mapper.rb:
```diff
@@ -0,0 +1,55 @@
+      module LDAP
+        # attribute mapper to redact sensitive keys
+        class AttributeMapper
+          SENSITIVE_KEYS = %w[userPassword unicodePwd lmPassword msDS-ManagedPassword ntPassword authPassword krbPrincipalKey sambaNTPassword sambaLMPassword].freeze
```
I think it is safer to have an allow list as opposed to a deny list. 

This will prevent any new sensitive keys from sneaking in.

scbjans committed:
chore: feedback

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,112 @@
+                end
+              end
+            rescue ::Net::LDAP::Error => e
+              span.add_attributes({
```
Thanks again for your review.

There is no good reason `record_exception` isn't used, fixed it in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/bce45ffd543dc723d72cf136556f8378378e39e3

According to [the RFC (4.1.2 String Types)](https://www.rfc-editor.org/rfc/rfc4511) this is always ISO10646 UTF-8 encoded, so it should be fine.

arielvalentin commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/attribute_mapper.rb:
```diff
@@ -0,0 +1,62 @@
+          SPAN_ATTRIBUTES =         %w[exception.message exception.stacktrace exception.type ldap.auth.method ldap.auth.username ldap.error.message
+                                       ldap.operation.type ldap.request.message ldap.response.status_code ldap.tree.base network.protocol.name
+                                       network.protocol.version network.transport peer.service server.address server.port].freeze
+          ALLOWED_KEYS = (LDAP_GENERAL_ATTRIBUTES + LDAP_OBJECT_ATTRIBUTES + SPAN_ATTRIBUTES).freeze
```
Have you considered using a `Set` to avoid a linear search?

arielvalentin commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/attribute_mapper.rb:
```diff
@@ -0,0 +1,62 @@
+                obj.map { |item| deep_map(item) }
+              end
+            when String
+              map_json(obj)
```
This looks like it will attempt to parse both JSON and String values but falling back to the raw string value. 

Are there specific fields that are JSON vs string literals? Can we avoid raising exceptions by checking that the string is a object type that needs to be parsed??

scbjans replied:
I'm not aware of a check where it could be validated if the String needs to be parsed.

The best I could think of, is checking if the String starts with `{` or `[`. That should avoid raising exceptions.

scbjans committed:
chore: use Set to avoid linear search, only map JSON when String needs to be parsed.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/attribute_mapper.rb:
```diff
@@ -0,0 +1,62 @@
+        # attribute mapper to redact keys which are not allowed
+        class AttributeMapper
+          LDAP_GENERAL_ATTRIBUTES = %w[attributes base filter ignore_server_caps left op operations paged_searches_supported replace right scope].freeze
+          LDAP_OBJECT_ATTRIBUTES  = %w[accountExpires codePage countryCode cn description displayName distinguishedName dn
```
yes, for these included fields, it will always be a JSON string.

Note: we've encountered an edge-case when someone searched with binary data in the filter. That was fixed in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/changes/621d92566d49d6595a8bc89288a6b340c034f272 but shouldn't even occur anymore, now we've switched to an allowlist for the attributes.

scbjans committed:
chore: fix bad merge, remove explicit bundler dependency

scbjans committed:
chore: add ostruct for net-ldap <0.20.0 and Ruby >=4

scbjans committed:
chore: do not appraise old net-ldap versions for ruby4-

scbjans committed:
chore: move ostruct to Gemfile instead of Appraisals

scbjans committed:
chore: also add logger to Gemfile

arielvalentin reviewed (commented):
One last question regarding exception handling. After that I think the 6 month long journey comes to an end.

scbjans committed:
chore: move ostruct to Appraisals. in_span already records exceptions, so do not record_exception again.

scbjans commented on instrumentation/net_ldap/lib/opentelemetry/instrumentation/net/ldap/instrumentation_service.rb:
```diff
@@ -0,0 +1,110 @@
+                  annotate_span_with_response(span, response) if response
+                end
+              end
+            rescue ::Net::LDAP::Error => e
```
indeed it does! Fixed in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/1587/commits/23069fd67b469e0b75e791cc3e3f951062d6a9e9

scbjans commented:
> One last question regarding exception handling. After that I think the 6 month long journey comes to an end.

Thank you (and @kaylareopelle and @thompson-tomo) so much for helping me get this in a better shape!

kaylareopelle commented:
Hi @scbjans, thank you for your patience with us on this PR. I'd love to get it merged in and released. There's a merge conflict with the release-please-config. Could you take a look at it?

scbjans commented:
> Hi @scbjans, thank you for your patience with us on this PR. I'd love to get it merged in and released. There's a merge conflict with the release-please-config. Could you take a look at it?

@kaylareopelle no worries, merge conflict is fixed. Failing test seem unrelated.

scbjans committed:
Merge branch 'main' into 669-add-net_ldap-instrumentation

scbjans committed:
Merge branch 'main' into 669-add-net_ldap-instrumentation
